flink-tests/src/test/java/org/apache/flink/test/checkpointing/StreamCheckpointNotifierITCase.java

```diff
@@ -0,0 +1,340 @@
+import org.apache.flink.util.Collector;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Random;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Integration test for the {@link CheckpointNotifier} interface. The test ensures that
+ * {@link CheckpointNotifier#notifyCheckpointComplete(long)} is called for some completed
+ * checkpoints, that it is called at most once for any checkpoint id and that it is not
+ * called for a deliberately failed checkpoint.
+ *
+ * <p>
+ * The topology tested here includes a number of {@link OneInputStreamOperator}s and a
+ * {@link TwoInputStreamOperator}.
+ *
+ * <p>
+ * Note that as a result of doing the checks on the task level there is no way to verify
+ * that the {@link CheckpointNotifier#notifyCheckpointComplete(long)} is called for every
+ * successfully completed checkpoint.
+ */
+@SuppressWarnings("serial")
+public class StreamCheckpointNotifierITCase extends StreamFaultToleranceTestBase {
+
+	final long NUM_LONGS = 10_000_000L;
+
+	/**
+	 * Runs the following program:
+	 *
+	 * <pre>
+	 *     [ (source)->(filter) ] -> [ (co-map) ] -> [ (map) ] -> [ (groupBy/reduce)->(sink) ]
+	 * </pre>
+	 */
+	@Override
+	public void testProgram(StreamExecutionEnvironment env) {
+
+		DataStream<Long> stream = env.addSource(new GeneratingSourceFunction(NUM_LONGS));
+
+		stream
+				// -------------- first vertex, chained to the src ----------------
+				.filter(new LongRichFilterFunction())
+
+				// -------------- second vertex, applying the co-map ----------------
+				.connect(stream).flatMap(new LeftIdentityCoRichFlatMapFunction())
+
+				// -------------- third vertex - the stateful one that also fails ----------------
+				.map(new IdentityMapFunction())
+				.startNewChain()
+
+				// -------------- fourth vertex - reducer and the sink ----------------
+				.groupBy(0)
+				.reduce(new OnceFailingReducer(NUM_LONGS))
+				.addSink(new SinkFunction<Tuple1<Long>>() {
+					@Override
+					public void invoke(Tuple1<Long> value) {
+						// do nothing
+					}
+				});
+	}
+
+	@Override
+	public void postSubmit() {
+		List[][] checkList = new List[][]{	GeneratingSourceFunction.completedCheckpoints,
+				IdentityMapFunction.completedCheckpoints,
+				LongRichFilterFunction.completedCheckpoints,
+				LeftIdentityCoRichFlatMapFunction.completedCheckpoints};
+
+		for(List[] parallelNotifications : checkList) {
+			for (int i = 0; i < PARALLELISM; i++){
+				List<Long> notifications = parallelNotifications[i];
+				assertTrue("No checkpoint notification was received.",
+						notifications.size() > 0);
+				assertFalse("Failure checkpoint was marked as completed.",
+						notifications.contains(OnceFailingReducer.failureCheckpointID));
+				assertTrue("Checkpoint notification was received multiple times",
+						notifications.size() == new HashSet<Long>(notifications).size());
```

Comment:
Maybe it would be also good to check whether we received anything after the failed checkpoint.